Created a buildController for PR purposes.

cppwfs opened the pull request:
It is called buildController-for-review.sh.

Once it meets Review standards we will delte the original buildController script and rename this one

cppwfs committed:
Created a buildController for PR purposes.

It is called buildController-for-review.sh.

Once it meets Review standards we will delte the original buildController script and rename this one

Signed-off-by: Glenn Renfro <grenfro@vmware.com>

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+#!/bin/bash
+
+# Set variables for repositories and workflows
```
We will want to move these to inputs w/ validation and usage. I am sure you already know this, just doing my due diligence...

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+
+    # retrieve the workflow id of the workflow yaml file
+    WF_ID=$(gh workflow list --repo "$WF_REPO" --json id,name,path,state --jq ".[] | select(.path | endswith(\"/$WF_PATH\")) | .id")
+    echo "WF_ID=$WF_ID"
```
Add validation that the WF_ID is not null

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+    gh workflow --repo "$WF_REPO" run "$WF_PATH" $WF_PARAM --ref "$WF_REF"
+    # Get its run ID
+    RUN_ID=
+    while [ "$RUN_ID" = "" ]; do
```
We don't want to spin forever as the RUN_ID will be available pretty quickly. Maybe 20s max??

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+      sleep 5
+      RUN_ID=$(gh run --repo "$WF_REPO" list --workflow=$WF_PATH --json status,workflowDatabaseId,headBranch,databaseId --jq ".[] | select(.headBranch == \"$WF_BRANCH\") | select(.workflowDatabaseId == $WF_ID) | {databaseId}" | jq -s '.[0].databaseId')
+      set -e
+      echo "RUN_ID=$RUN_ID"
```
If `RUN_ID` is `""` probably don't want to echo it

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+      echo "RUN_ID=$RUN_ID"
+    done
+    # wait for completion and check every 10s
+    gh run --repo "$WF_REPO" watch $RUN_ID -i 10
```
Does `watch` have a max wait time? I can see this taking a decent amount of time but we should constrain it somehow.

cppwfs replied:
Unfortunately no 😞

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+    gh run --repo "$WF_REPO" watch $RUN_ID -i 10
+    # log output
+    set -e
+    gh run --repo "$WF_REPO" view $RUN_ID --log --exit-status
```
I think we can delete this line as its done in next line

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,60 @@
+}
+
+# Trigger workflow in Repository A
+echo "Launching workflow in $REPO_A..."
```
I would include `$WF_PATH_A $BRANCH_A`

onobc reviewed (changes requested):
Looking good @cppwfs . Some comments around  max wait times and improved logging.. Other than that LGTM

cppwfs committed:
Updated buildCCOntroller-for-review.sh based on code review

* Added docs and instructions if no parameters are passed in
* renamed buildController.sh to workflow_runner.

Once PR is accepted.  I'll delete workflow_runner.sh and rename buildController-for-review.sh to workflow_runner.sh

Signed-off-by: Glenn Renfro <grenfro@vmware.com>

cppwfs committed:
Update BuildController Yaml to call workflow runner script

onobc commented:
> It is called buildController-for-review.sh.
> 
> Once it meets Review standards we will delte the original buildController script and rename this one

FWIW I like `run-workflow.sh`

cppwfs committed:
Apply code review changes

* Renamed builder to run-workflow.sh
* update comment to show that multiple workflows can be executed
* Refactored process_workflow_input_tuple to control the parsing of tuples vs just setting the array value.

Signed-off-by: Glenn Renfro <grenfro@vmware.com>

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,97 @@
+
+# if no arguments provided provide instructions
+if [[ "$#" == 0 ]]; then
+  echo "This Script will execute the specified group of workflows passed in via command line args"
```
`s/This Script will execute/Executes`

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,97 @@
+
+  # Iterate over each element in the current comma-separated list
+  for element in "${elements[@]}"; do
+    process_comma_delimited_list "$element"
```
Nit but I think if a function name like `process_workflow_input _tuple` may be a bit more readable

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,97 @@
+  for element in "${elements[@]}"; do
+    process_comma_delimited_list "$element"
+  done
+    arraySize=${#workflow[@]}
```
All below here needs left indent

onobc commented on scripts/buildController-for-review.sh:
```diff
@@ -0,0 +1,97 @@
+  for element in "${elements[@]}"; do
+    process_comma_delimited_list "$element"
+  done
+    arraySize=${#workflow[@]}
```
I thought this needed to be indented but now I am confused as to what is happennig

cppwfs replied:
It shouldn't be indented.
Lines 83-96 will be shifted to the left one tab.

cppwfs replied:
The main is a loop that goes through each of the workflow groups.
For each workflow group it extracts the repo,branch,workflow.yml file for the workflow group.   Then runs the workflow based on this information.

I can add a comment.

onobc replied:
Bash sucks! I get it now. I made some comment on method and var renames that will make it a bit easier to follow.

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+workflow=()
+# Function to extract each item from comma delimited list
+# Parameter comma delimited list of workflow args
+process_comma_delimited_list() {
```
I would move this down closer to where its used

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+  exit 0
+fi
+
+workflow=()
```
Maybe naming this `workflowInputTuple` would make it easier to follow?

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+fi
+
+workflow=()
+# Function to extract each item from comma delimited list
```
> Function to extract a workflow input tuple into 3 elements in the output variable `workflowInputTuple`

Wdyt? The mystery of bash in/out otherwise took me for a ride

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+  # Iterate over each element in the current comma-separated list
+  for element in "${elements[@]}"; do
+    process_comma_delimited_list "$element"
+  done
```
Is this `done` supposed to be here?

cppwfs replied:
Yes.  This where we extract the repo, branch, workflow.yml file data from each element.
For example `cppwfs/toupperish,main,foo.yml spring-cloud/spring-cloud-dataflow,main,bar.yml`
Will produce 2 elements.  Each element will have the repo, branch,and workflowyaml.

onobc replied:
Oh this is that terrible IFS csv that bash uses.... I get it now. Maybe rename `elements` and `element` as that is a very generic name. 

What I would do is abstract everything from 76-87 in a function called `parse_input_tuple_into_workflow_coordinates`. really we are looping through N tuple-csv-strings and for each one we want 3 fields (thought is "don't care what nasty things you have to do to get it just give me the output" :)

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+  for element in "${elements[@]}"; do
+    process_comma_delimited_list "$element"
+  done
+    arraySize=${#workflow[@]}
```
I think moving this validation into the `process_workflow_input_tuple` method would simplify things

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,97 @@
+# if no arguments provided provide instructions
+if [[ "$#" == 0 ]]; then
+  echo "This Script will execute/executes the specified group of workflows passed in via command line args"
+  echo "A workflow is specified in a comma delimited format as follows: REPOSITORY,BRANCH,NAME_OF_THE_WORKFLOW_YAML_FILE"
```
I would replace this with a "usage()" function that was a little more robotic and less personal. Something like:

```
Executes one or workflows in serial fail-fast fashion.

usage: ./run-workflow.sh [workflow-tuple]*
    where [workflow-tuple] is a csv string of workflow coordinates in the form
    "repo,branch,workflow" (e.g. "my-repo,main,my-pr.yml")

```
Wdyt?

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,92 @@
+  echo ""
+  echo "usage: ./run-workflow.sh [workflow-tuple]*"
+  echo "    where [workflow-tuple] is a csv string of workflow coordinates in the form"
+  echo '    "repo,branch,workflow" (e.g. "my-repo,main,my-pr.yml")'
```
Only in bash do you have one string line starting w/ `"` and the next starting w/ `'` (and I know exactly why you had to do it).

cppwfs replied:
Don't tell my mom :-)

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,92 @@
+
+    # retrieve the workflow id of the workflow yaml file
+    WF_ID=$(gh workflow list --repo "$WF_REPO" --json id,name,path,state --jq ".[] | select(.path | endswith(\"/$WF_PATH\")) | .id")
+    if [ -z "${WF_ID}" ]; then
```
The missing piece always w/ these bash script (for me anyways) is how to test them. Yeh, the answer is manually. That being said, the 1000 different ways to check if an arg is empty or null or ... I would definitely manually test this by setting WF_ID ="" and WF_ID= and making sure it does what you think. I think it does but I never trust these operators unless I see it.

cppwfs replied:
Kicked the tires and It checks for null and empty successfully.

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,92 @@
+# Function to extract a workflow input tuple into 3 elements in the output variable workflowInputTuple
+process_workflow_input_tuple() {
+    # Set IFS to comma to split comma-delimited lists to obtain workflow information
+    IFS=',' read -ra elements <<< "$arg"
```
Ahh yes, the magic of "what in the hell is 'arg' and where did it come from? AHh yes, bash I hate you.

I prefer to pass the arg in such as this:
```
process_workflow_input_tuple() {
    inputTuple = $1
   ...
}
```
and to call it....
```
process_workflow_input_tuple arg
```

Wdyt?

onobc replied:
Moving this to the function definitely cleans up the main method.

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,92 @@
+
+ process_workflow_input_tuple
+
+  repository=${workflowInputTuple[0]}
```
**[TinyNit]** indentation off below

onobc commented on scripts/run-workflow.sh:
```diff
@@ -0,0 +1,92 @@
+  echo "Launching workflow $workflow_path for repository $repository for branch $branch"
+  launch_workflow_and_wait "$repository" "$workflow_path" "$branch"
+
+  workflowInputTuple=()
```
Oh my bash!!!  hehe

I think we can move this to line 65. I know it has to be defined outside of the function at least once (which it is on 62). That way it keeps that to an internal concern.

cppwfs committed:
Update based on code review

Signed-off-by: Glenn Renfro <grenfro@vmware.com>
